hicommonwealth/commonwealth
# ABI Clean-up

## Link to Issue
Closes: #5066

## Description of Changes
- Created the `hashAbi` function which defines the ABI hashing strategy.
- Added a new `abi_hash` column to the `ContractAbis` table. Populated the column for existing ABIs.
- Removed the unique constraint on the `abi` column and recreated it on the `abi_hash` column.
- Added a `CHECK` constraint on the `abi` column which verifies that the value in the column is an array (prevents stringified ABIs).
- Duplicated the `CHECK` constraint as a validation function on the Sequelize model for visibility.
- Added a `beforeValidate` hook on the Sequelize model to generate the hash if it is not already given.
- Updated existing queries to the `ContractAbis` table to use the new `abi_hash` column when searching for an existing ABI.
- Deleted the `createContractAbi` function and route since it is entirely unused. @ianrowan confirmed this deletion is okay.
- Modified `createContract` so that if an old contract exists without an ABI but an ABI is provided in the request data, the ABI is found or created and linked to the old contract.

## "How We Fixed It"

- Introduced an ABI hashing method that is used to ensure uploaded ABIs are unique.
- Added Sequelize and database level validation to ensure uploaded ABIs have the proper format.

## Test Plan
- Become admin of any EVM community e.g. dYdX
- Go to the 'Create action template' page.
- Open networks tab of chrome dev tools
- Copy the address of the existing linked contract.
- Delete the contract.
- Click on 'Add new contract' button
- Paste the contract address.
- The ABI should load automatically.
- Click the add button to save the contract.
  - There should be a successful response to `/contract` and `/template`

## Deployment Plan

1. 

## Other Considerations

-
